### ISSUE TITLE
Failing to guess a single MS-apostrophe

### ISSUE THREAD

**Issue by shompol** [chardet/chardet#53]
I have a page of text in ASCII with a single Microsoft-apostrophe `chr(8217)` detected as **ISO-8859-2**. 

``` python
#1. Create problematic sample
>>> s = 'today' + chr(8217) + 's research'
>>> s
'today’s research'
>>> b = s.encode('windows-1252')
>>> b
b'today\x92s research'

#2. Attempt to decode it
>>> chardet.detect(b)
{'encoding': 'ISO-8859-2', 'confidence': 0.8060609643099236}
>>> b.decode('ISO-8859-2')
'today\x92s research'

#3. Now try the correct encoding
>>> b.decode('windows-1252')
'today’s research'
```

This text is very typical of anything created using a Microsoft editor. Furthermore, latest version of Firefox detects it correctly. I am using Python 3.3. Any help is appreciated.

**Comment by helour** [comment]
The original mozilla based algorithm detects singlebyte charset only via the **letters** bigrams = two-letters sequences without any symbols (e.g. apostrophes, ...)

**Comment by helour** [comment]
1. Don't try to detect the encoding from two words. You need at least 1024 characters.
2. The Hungarian (iso-8859-2) language prober with its language model is bad and it leads to wrong results.

**Comment by shompol** [comment]
The text in question is 3240 characters long, all characters are ASCII except for the ill-fated apostrophe. The sample above is sufficient to replicate the issue.

I run it with `chardet.constants._debug=1` and see that `windows-1252` is not even attempted. Documentation says that `windows-1252` is only attempted as the last resort. This is very wrong in my case since I am in North America. I guess the best way to proceed is to override a few methods...

**Comment by dan-blanchard** [comment]
@shompol I've been a bit behind on things lately since this is just a side project for me and I have a newborn son, but I plan to make a release "soon" (probably in the next month) that either disables the Hungarian prober (as is currently the case in `master`), or switches to the retrained version in #52.  I would suggest giving the version in master a try and seeing if you still see the same issue.

**Comment by dan-blanchard** [comment]
Also, Firefox detects things correctly because they disabled their Hungarian prober a while ago.  I didn't realize that until a month or two ago.

**Comment by shompol** [comment]
Thank you Dan, the master version works much better:
```python
>>> import chardet
>>> s = 'today' + chr(8217) + 's research'
>>> b = s.encode('windows-1252')
>>> chardet.detect(b)
{'encoding': 'windows-1252', 'confidence': 0.73}
```
The original 3KB text detection works as well.

May I inquire why the Latin-1 confidence is artificially reduced by 0.73? It seems a little odd to me, especially since Latin-1 covers 99% of all texts I work with in US.

Congratulations with the son! :)

**Comment by sigmavirus24** [comment]
@dan-blanchard feel free to assign me a bug for a new release.

**Comment by helour** [comment]
BTW, I'm working on the new (mixed = more than one) language model for latin1 (cp1252) prober, which replaces old one (which is based on very "strange heuristic" for me).  The new model will be based on the statistic probability like other "normal" single byte probers.

**Comment by helour** [comment]
BTW, I'm working on the new (mixed = more than one) language model for latin1 (cp1252) prober, which replaces old one (which is based on very "strange heuristic" for me).  The new model will be based on the statistic probability like other "normal" single byte probers.

**Comment by shompol** [comment]
Just a note that apparently chardet is also used by BeautifulSoup, and it got fixed too:
```python
>>> s = 'today' + chr(8217) + 's research'
>>> b = s.encode('windows-1252')
>>> from bs4 import BeautifulSoup
>>> soup = BeautifulSoup(b)
>>> soup
<html><head></head><body>today’s research</body></html>
>>> soup.original_encoding
'windows-1252'
```
* - BeautifulSoup implicitly uses an HTML parser, in my case it is `html5lib`

**Comment by dan-blanchard** [comment]
> Just a note that apparently chardet is also used by BeautifulSoup, and it got fixed too

BeautifulSoup uses cChardet if it's available, and only uses chardet if it's available.  Unfortunately, cChardet and chardet have very different detection characteristics these days.  We can handle more encodings than cChardet, and for some of them we're more accurate, but they're much faster (and already had the Hungarian detector disabled).

**Comment by helour** [comment]
Today's result with "non destructive" filtering (= tags, urls, html code numbers, digits, some symbols, empty lines, spaces removing) and with the hungarian langmodel and tests enabled:
<code>
Ran 416 tests in 56.891s
FAILED (failures=**2**)
</code>
what gives less than 0.5% failures. It is really good accuracy. All changes are in the my own fork.

**Comment by helour** [comment]
@shompol: *May I inquire why the Latin-1 confidence is artificially reduced by 0.73?* 
The latin1 prober is based on the "heuristic table" with letters types/categories (small ascii, capital ascii, small ascii with accute, ...) and it is not normal/common language model. The values >0.73 causes that the latin1 prober wins before other common lang models which can leads to wrong detection. There are many languages in the Latin1 group and the better solution is to create "heuristic table" instead of many language models.

**Comment by helour** [comment]
My own "milestone" was achieved when I've found solution of the first failure with greek **Ά**:
```
Ran 416 tests in 62.030s

OK
```

**Comment by dan-blanchard** [comment]
@helour That's fantastic news! I'm sorry I've been slow on the uptake of this, but I'll definitely check this out soon.  Either this weekend or next week.

**Comment by dan-blanchard** [comment]
Fixed in 3.0.
